subsys/net/ip/net_context.c

```diff
@@ -1814,29 +1814,27 @@ NET_CONN_CB(tcp_syn_rcvd)
 		}
 
 		/* Swap the newly-created TCP states with the one that
-		 * was used to establish this connection. The new TCP
+		 * was used to establish this connection. The old TCP
 		 * must be listening to accept other connections.
 		 */
-		tmp_tcp = new_context->tcp;
-		tmp_tcp->accept_cb = tcp->accept_cb;
-		tcp->accept_cb = NULL;
-		new_context->tcp = tcp;
+		new_tcp = new_context->tcp;
+		new_tcp->accept_cb = tcp->accept_cb;
 		copy_pool_vars(new_context, context);
-		context->tcp = tmp_tcp;
 
-		tcp->context = new_context;
-		tmp_tcp->context = context;
+		net_tcp_change_state(tcp, NET_TCP_LISTEN);
 
-		net_tcp_change_state(tmp_tcp, NET_TCP_LISTEN);
+		/* We cannot use net_tcp_change_state() here as that will
+		 * check the state transitions. So set the state directly.
+		 */
+		new_context->tcp->state = NET_TCP_ESTABLISHED;
 
-		net_tcp_change_state(new_context->tcp, NET_TCP_ESTABLISHED);
 		net_context_set_state(new_context, NET_CONTEXT_CONNECTED);
 
-		context->tcp->accept_cb(new_context,
-					&new_context->remote,
-					addrlen,
-					0,
-					context->user_data);
+		new_context->tcp->accept_cb(new_context,
```

Comment:
Well, accept_cb should be called on listening ctx, not on newly accepted one.

Reply:
I will investigate this more.